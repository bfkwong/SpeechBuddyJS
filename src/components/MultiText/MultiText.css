.dropdown-container {
  margin: 20px 0px 20px 0px;
}

.MultiTxt__aside {
  padding-top: 20px;
  text-align: left;
}

.MultiTxt__analysisContainer {
  margin-top: 20px;
}
