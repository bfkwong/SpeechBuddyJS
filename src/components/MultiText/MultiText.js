import React, { useCallback, useState } from "react";
import { useDropzone } from "react-dropzone";
import { Container, Row, Col, Button, Table } from "react-bootstrap";
import styled from "styled-components";
import * as use from "@tensorflow-models/universal-sentence-encoder";
import * as similarity from "compute-cosine-similarity";
import * as natural from "natural";
import "./MultiText.css";

const StyledDiv = styled.div`
  background-color: #d1f0f0;
  color: #23686c;
  flex: 1;
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 50px;
`;

function MultiText() {
  const [analysis, setAnalysis] = useState([]);
  const [fileContents, setFileContents] = useState([]);

  const onDrop = useCallback((acceptedFiles) => {
    acceptedFiles.forEach((file) => {
      const reader = new FileReader();
      reader.onload = () => {
        const string = reader.result;
        fileContents.push(string);
        setFileContents(fileContents);
      };
      reader.readAsText(file);
    });
  }, []);

  const { acceptedFiles, getRootProps, getInputProps } = useDropzone({
    onDrop
  });

  const filesUpload = acceptedFiles.map((file, i) => file.path);

  const runAnalysis = async () => {
    let model = await use.load();
    let embeddings = await model.embed(fileContents);
    let values = embeddings.arraySync();
    let arr = Array.from(values);

    let output = [];
    for (let i = 0; i < arr.length - 1; i++) {
      for (let j = i + 1; j < arr.length; j++) {
        let csd = similarity(arr[i], arr[j]);
        let jwd = natural.JaroWinklerDistance(fileContents[i], fileContents[j]);
        output.push([filesUpload[i], filesUpload[j], csd, jwd]);
      }
    }
    setAnalysis(output);
  };

  return (
    <div className="TopLevelDisplay">
      <h1>Multi Text Analysis</h1>
      <div className="dropdown-container">
        <section className="container">
          <StyledDiv {...getRootProps({ className: "dropzone" })}>
            <input {...getInputProps()} />
            <p>Drag 'n' drop some files here, or click to select files</p>
          </StyledDiv>
          <aside style={{ paddingTop: "20px", textAlign: "left" }}>
            <h4>Uploaded Files</h4>
            <Container>
              <Row>
                {filesUpload.map((f, i) => (
                  <Col md={2} key={i}>
                    <p>{f}</p>
                  </Col>
                ))}
              </Row>
            </Container>
          </aside>
        </section>
      </div>
      <Container>
        <Row className="justify-content-md-center">
          <Col xs={12} md={4}>
            <Button onClick={() => runAnalysis()}>Analyze</Button>
          </Col>
        </Row>
      </Container>
      {analysis.length > 0 && (
        <Container style={{ marginTop: "20px" }}>
          <Row className="justify-content-md-center">
            <Col>
              <Table>
                <thead>
                  <tr>
                    <th>Text</th>
                    <th>Text</th>
                    <th>Cosine Similarity</th>
                    <th>Jaro Winkler Similarity</th>
                  </tr>
                </thead>
                <tbody>
                  {analysis.map((a, i) => (
                    <tr key={i}>
                      <td>{a[0]}</td>
                      <td>{a[1]}</td>
                      <td>{a[2]}</td>
                      <td>{a[3]}</td>
                    </tr>
                  ))}
                </tbody>
              </Table>
            </Col>
          </Row>
        </Container>
      )}
    </div>
  );
}

export default MultiText;
