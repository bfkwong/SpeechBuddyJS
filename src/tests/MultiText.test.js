//Daniel Tisdale

import React from "react";
import { render } from "@testing-library/react";
import MultiText from "../components/MultiText/MultiText";
import renderer from "react-test-renderer";
import { unmountComponentAtNode } from "react-dom";

describe("MultiText Component", () => {
  let container;

  beforeEach(() => {
    container = document.createElement("div");
    document.body.appendChild(container);
  });

  afterEach(() => {
    if (container.hasChildNodes()) {
      unmountComponentAtNode(container);
    }
    container.remove();
    container = null;
  });

  test("MultiText renders without failure", () => {
    render(<MultiText />, container);
  });

  test("MultiText snapshot test", () => {
    const appTree = renderer.create(<MultiText />).toJSON();
    expect(appTree).toMatchSnapshot();
  });
});
