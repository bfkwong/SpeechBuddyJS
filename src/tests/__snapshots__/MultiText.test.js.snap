// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiText Component MultiText snapshot test 1`] = `
<div
  className="TopLevelDisplay"
>
  <h1>
    Multi Text Analysis
  </h1>
  <div
    className="dropdown-container"
  >
    <section
      className="container"
    >
      <div
        className="sc-AxjAm edvKor dropzone"
        onBlur={[Function]}
        onClick={[Function]}
        onDragEnter={[Function]}
        onDragLeave={[Function]}
        onDragOver={[Function]}
        onDrop={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        tabIndex={0}
      >
        <input
          autoComplete="off"
          multiple={true}
          onChange={[Function]}
          onClick={[Function]}
          style={
            Object {
              "display": "none",
            }
          }
          tabIndex={-1}
          type="file"
        />
        <p>
          Drag 'n' drop some files here, or click to select files
        </p>
      </div>
      <aside
        className="MultiTxt__aside"
      >
        <h4>
          Uploaded Files
        </h4>
        <ul />
      </aside>
    </section>
  </div>
  <div
    className="container"
  >
    <div
      className="justify-content-md-center row"
    >
      <div
        className="col-md-4 col-12"
      >
        <button
          className="btn btn-primary"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Analyze
        </button>
      </div>
    </div>
  </div>
</div>
`;
